rfd/0095-teleport-key-agent-extension.md

```diff
@@ -0,0 +1,129 @@
+---
+authors: Brian Joerger (bjoerger@goteleport.com)
+state: draft
+---
+
+# RFD 95 - Teleport Key Agent Extension
+
+## Required Approvers (TBD)
+
+## What
+
+Extend Teleport Key Agent to allow users make `tsh` requests from within a remote session started with `tsh ssh -A`.
+
+## Why
+
+It's already possible to `ssh` from a remote session in this way, but `tsh` requires more than just an SSH certificate and private key to start a session. At the minimum, `tsh` also needs a TLS certificate and proxy address to form connections to a Teleport proxy. To avoid using `tsh --insecure`, the remote session will also need access to a list of SSH and TLS CA certificates.
+
+Thus, we need a secure protocol for sharing certificates and other relevant data on the local host with the remote host. We can lean on the [extensible](https://datatracker.ietf.org/doc/html/rfc4251#section-4.2) SSH Agent protocol for this, by implementing custom extensions for retrieving TLS certificates, CA certificates, and even perform per-session MFA signatures with the Teleport Key Agent.
+
+## Details
+
+### Teleport Key Agent Changes
+
+We will add new [SSH Agent extensions](https://datatracker.ietf.org/doc/html/draft-miller-ssh-agent#section-4.7) to forward additional `tsh` keystore items and functions, namely TLS certificates, CA certificates, known hosts, and per-session MFA reissue. Basic SSH Agent functionality will not be impacted by adding these extensions, so existing integrations will remain unchanged.
+
+To support these extensions, `tsh` will need to be updated to extend the current Teleport Key Agent and add a new Teleport Key Agent client. The Teleport Key Agent client can be used by `tsh` calls on a remote host to send SSH Agent requests to the Teleport Key Agent running on the local host.
+
+#### SSH Agent Extensions
+
+**Extension Structure:**
+
+Extension requests consist of:
+
+        byte            SSH_AGENTC_EXTENSION
+        string          extension type (type@domain)
+        byte[]          extension contents
+
+An Extension response can be any custom message, but failures should result in `SSH_AGENT_EXTENSION_FAILURE`. Unsupported extensions should result in an `SSH_AGENT_FAILURE` response to differentiate from actual extension failures.
+
+**TLS Certificate Extension:**
```

Comment:
This will make the protocol very chatty and will require a lot of roundtrips. (3 extra for each connection?) Instead, can we pack all the data in combined requests? E.g. cert, CAs and known-hosts in one request/response?

Reply:
Good point, I updated it to a single extension.